### PR TITLE
[next] feat(messaging-amqp)

### PR DESCRIPTION
**Pull request by pierissimo**
# Changes included
This PR is the first attempt to start making DaVinci a generic App framework, also for headless applications.
Specifically, it includes a new Amqp module that allow a DaVinci app to subscribe and publish on the amqp protocol.

- Added decorators: channelParam, message, payload, subscribe
- Implemented ability to specify a behaviour for shutting down gracefully the amqp service
- Reusing channels by default, to avoid stressing the server

### CODE REVIEW

File: packages/messaging-amqp/src/AmqpModule.ts [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+	};
+}
+
+export class AmqpModule extends Module {
+	app: App;
+	private logger: Logger;
+	private options: AmqpModuleOptions;
+	private connection: AmqpConnectionManager;
+	private subscriptions: Array<Subscription>;
+	private subscriptionHash: Record<string, Subscription> = {};
```
Comment (joseandrespg) [inline packages/messaging-amqp/src/AmqpModule.ts:46]: Should we move to use `Map` instead of `Object` for these dictionaries?

**Review by joseandrespg — COMMENTED** [review]
Would this module also handle publishing?

File: packages/messaging-amqp/src/AmqpModule.ts (hunk 1 above)
Comment (pierissimo) [reply packages/messaging-amqp/src/AmqpModule.ts:46]: 👍

**Comment by pierissimo** [comment]
@joseandrespg yes, currently you can use the `channel()` decorator in the controller method, to inject the channel object and publish with it. I'll add an example to make it clearer.

File: packages/messaging-amqp/src/AmqpModule.ts [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+
+	getModuleId() {
+		return ['messaging', 'amqp'];
+	}
+
+	onRegister(app: App) {
+		this.app = app;
+		const connectionOptions =
+			typeof this.options.connection === 'object' ? this.options.connection : { url: this.options.connection };
+		this.connection = amqpConnectionManager.connect(connectionOptions);
```
Comment (juanjoseruiz) [inline packages/messaging-amqp/src/AmqpModule.ts:67]: is there any way to specify heartbeatIntervalInSeconds / reconnectTimeInSeconds ? I think we are using those ones in messaging lib.

**Review by juanjoseruiz — COMMENTED** [review]
looks good to me, pretty straightforward to use. 

What's the idea of supporting this at DaVinci level? Asking because we do have some custom logic in our current messaging lib so I am not sure if the idea is to replace it (for APIs) or if just offer a simpler & straightforward option.

**Comment by pierissimo** [comment]
@juanjoseruiz 
The idea is to use DaVinci as the standard framework both for APIs and headless services.
But I agree, we have a custom and opinionated message library with a bunch of specific functionalities for our team.
For that reason I added [the ability into the module](https://github.com/HPInc/davinci/pull/143/commits/42c1df6ed67e73077782e48219844c141c3a79f3#diff-13e58c3b056805218397a04f5c6cd0094093da53d3292af269949df5596c4d04R75-R77), to accept a custom ChannelManager implementation so that the subscription and channel logic can be handling externally if needed

**Comment by juanjoseruiz** [comment]
> @juanjoseruiz The idea is to use DaVinci as the standard framework both for APIs and headless services. But I agree, we have a custom and opinionated message library with a bunch of specific functionalities for our team. For that reason I added [the ability into the module](https://github.com/HPInc/davinci/pull/143/commits/42c1df6ed67e73077782e48219844c141c3a79f3#diff-13e58c3b056805218397a04f5c6cd0094093da53d3292af269949df5596c4d04R75-R77), to accept a custom ChannelManager implementation so that the subscription and channel logic can be handling externally if needed. It's currently limited to the subscriptions, but I'll add the functionality to publish too

Sounds good, thanks! Yes I see it's very flexible so we can extend it to match custom logic if needed.

Are you also planning to do the same for SQS?

**Comment by pierissimo** [comment]
Yes, SQS will be next!

**Pull request merged**